src/clients/dataset-client.ts

```diff
@@ -52,56 +56,42 @@ export class ExtDatasetClient<T extends DatasetItem> extends DatasetClient<T> im
     }
 
     async* greedyIterate(options: GreedyIterateOptions = {}): AsyncGenerator<T, void, void> {
-        const { pageSize = 100, itemsThreshold = 100, pollIntervalSecs = 10, ...listItemOptions } = options;
-        this.customLogger.info('Greedily iterating Dataset', { pageSize }, { url: this.url });
+        const { pollIntervalSecs = 10, ...iterateOptions } = options;
+        this.customLogger.info('Greedily iterating Dataset', { pageSize: iterateOptions.pageSize }, { url: this.url });
 
-        let readItemsCount = 0;
+        let currentOffset = iterateOptions.offset ?? 0;
 
-        let dataset: Dataset | undefined;
-        let run: ActorRun | undefined;
+        const runId = (await this.get())?.actRunId;
 
-        // TODO: breaking change - remove itemsThreshold and just listItems at every iteration
-        while (true) {
-            dataset = await this.get();
-            if (!dataset || !dataset.actRunId) {
-                this.customLogger.error('Error getting Dataset while iterating greedily', { id: this.id });
-                return;
-            }
-
-            run = await this.apifyClient.run(dataset.actRunId).get();
-            if (!run) {
-                this.customLogger.error('Error getting Run while iterating Dataset greedily', { id: this.id });
-                return;
-            }
+        let runStatus = runId ? (await this.apifyClient.run(runId).get())?.status : undefined;
 
-            if (run.status !== 'READY' && run.status !== 'RUNNING') {
-                break;
-            }
-
-            if (dataset.itemCount >= readItemsCount + itemsThreshold) {
-                const itemList = await this.superClient.listItems({ ...listItemOptions, offset: readItemsCount, limit: pageSize });
-                readItemsCount += itemList.count;
-                for (const item of itemList.items) {
+        if (runId) {
+            while (runStatus && ['READY', 'RUNNING'].includes(runStatus)) {
+                const datasetIterator = this.iterate({ ...iterateOptions, offset: currentOffset });
+                for await (const item of datasetIterator) {
+                    currentOffset++;
```

Comment:
I don't like this, and postponed adding my review till I had let it settle a bit. I ended up having to do pretty much the same when using the library to do "reasonable segmentation". I think this shows that our abstraction tries to paper over leaky issues - how many elements we got from each successive call to the API. I think it would be better if we just returned `AsyncGenerator<T[], void, void>`, only yielding if there actually were any elements. I think `iterate` should have the same signature, even though it may often be used without a page size such it only actually yields once.

Reply:
I think that would simplify both this code, and most (if not all) usages